backend/queue/queue_test.go

```diff
@@ -270,3 +202,202 @@ func TestNackExpired(t *testing.T) {
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	// At this point q1, q2 and q3 all have items waiting
+	require.NoError(t, q1.Enqueue(ctx, "foobar"))
+
+	// Truncate the member list to be only q1
+	oldMembers := memberListResponse.Members
+	memberListResponse.Members = oldMembers[:1]
+
+	ctx, cancel = context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	// After the next enqueue operation, the values for q2 and q3 will be removed
+	require.NoError(t, q1.Enqueue(ctx, "barbaz"))
+
+	// Restore the member list to its original state
+	memberListResponse.Members = oldMembers
+
+	ctx, cancel = context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	require.NoError(t, q1.Enqueue(ctx, "rubber ducky"))
+
+	// q1 expects to get three values
+	for _, exp := range []string{"foobar", "barbaz", "rubber ducky"} {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+		defer cancel()
+		item, err := q1.Dequeue(ctx)
+		require.NoError(t, err)
+		assert.Equal(t, exp, item.Value())
+	}
+
+	// q2 and q3 expect only a single value
+	for _, q := range []*Queue{q2, q3} {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+		defer cancel()
+
+		item, err := q.Dequeue(ctx)
+		require.NoError(t, err)
+		assert.Equal(t, "rubber ducky", item.Value())
+	}
+}
+
+func TestDequeueParallel(t *testing.T) {
+	t.Parallel()
+	e, cleanup := etcd.NewTestEtcd(t)
+	defer cleanup()
+	client, err := e.NewClient()
+	defer client.Close()
+	require.NoError(t, err)
+	queue := New("testparallel", client, e.BackendID())
+	items := map[string]struct{}{
+		"hello":   struct{}{},
+		"there":   struct{}{},
+		"world":   struct{}{},
+		"asdf":    struct{}{},
+		"fjdksl":  struct{}{},
+		"lalalal": struct{}{},
+	}
+	var mu sync.Mutex
+	var wg sync.WaitGroup
+	wg.Add(len(items))
+	var errEnqueue error
+	for item := range items {
+		go func(item string) {
+			defer wg.Done()
+			// Prevent data races when inspecting the error
+			mu.Lock()
+			defer mu.Unlock()
+			if err := queue.Enqueue(context.Background(), item); err != nil {
+				errEnqueue = err
+			}
+		}(item)
+	}
+	wg.Wait()
+	// Make sure we didn't encountered any error when adding items to the queue.
+	// If we had multiple errors, only the last one is saved
+	require.NoError(t, errEnqueue)
+	results := make(map[string]struct{})
+	var errDequeue error
+	wg.Add(len(items))
+	for range items {
+		go func() {
+			defer wg.Done()
+			item, err := queue.Dequeue(context.Background())
+			// Prevent data races when inspecting the error or the result
+			mu.Lock()
+			defer mu.Unlock()
+			if err != nil {
+				errDequeue = err
+				return
+			}
+			results[item.Value()] = struct{}{}
+		}()
+	}
+	wg.Wait()
+	// Make sure we didn't encountered any error while dequeuing items from the
+	// queue. If we had multiple errors, only the last one is saved
```

Comment:
This is a true nitpick if there ever was one, but "didn't encountered" -> "don't encounter" or something similar.